subscribe.go

```diff
@@ -0,0 +1,82 @@
+package bps
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"sync"
+)
+
+var (
+	subReg   = map[string]SubscriberFactory{}
+	subRegMu sync.Mutex
+)
+
+// SubscribeOptions defines subscribe options.
+type SubscribeOptions struct {
+	// BatchSize holds desired message batch size.
+	// Default: 1.
+	BatchSize int
+}
+
+// Normalize normalises subscriber options by applying defaults.
+func (o *SubscribeOptions) Normalize() *SubscribeOptions {
+	if o == nil {
+		o = &SubscribeOptions{}
+	}
+	if o.BatchSize == 0 {
+		o.BatchSize = 1
+	}
+	return o
+}
+
+// ----------------------------------------------------------------------------
+
+// Subscriber defines the main subscriber interface.
+type Subscriber interface {
+	// Subscribe subscribes for topic messages and blocks till context is cancelled or error occurs.
+	//
+	// Handler is provided an atomic message batch (entire batch either succeeds or fails).
+	// Batch is guaranteed to contain at least 1 message.
+	Subscribe(
+		ctx context.Context,
+		topic string,
+		handler func([]Message) error,
+		opts *SubscribeOptions,
+	) error
+	// Close closes the subscriber connection.
+	Close() error
+}
+
+// NewSubscriber inits to a subscriber via URL.
+//
+//   sub, err := bps.NewSubscriber(context.TODO(), "kafka://10.0.0.1:9092,10.0.0.2:9092,10.0.0.3:9092/namespace")
```

Comment:
can we use the same URL pattern for producers and consumers?

Reply:
@dim exactly what I'm doing, check this: https://github.com/bsm/bps/pull/3